### PR TITLE
update for dealing with big dataset (size_t, remove conversion from distance to matrix) and verbose for running details, etc.

### PR DESCRIPTION
**Pull request by Xiaojieqiu**
I have made the following changes to enable the package to deal with big dataset (tested on about 70k samples): 
1. replace int with size_t to declare variable in Rcpp 
2. remove the conversion from dist object to matrix in findClusters.densityCluster 
3. added a findDistValueByRowColInd function writen in Rcpp to accelerate the access of elements in the dist object mentioned above 
4. add verbose for some functions for printing running information and enable findClusters.densityCluster to accept a vector of index as the density peaks 
5. added new code in test.that for testing the findDistValueByRowColInd function

### CODE REVIEW

**Comment by thomasp85** [comment]
Thanks for this - I've been a bit slow due to some life-transition but I will go through it today

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
+MD5
```
Comment (thomasp85) [inline .gitignore]: Can you avoid changing non-r files?

File: DESCRIPTION [hunk 2]
```diff
@@ -12,9 +12,13 @@ Description: An implementation of the clustering algorithm described by Alex
...
-Imports:
-    Rcpp
+Imports: Rcpp
 RoxygenNote: 5.0.1
+NeedsCompilation: yes
+Packaged: 2016-02-06 12:28:04 UTC; Thomas
+Author: Thomas Lin Pedersen [aut, cre],
+  Sean Hughes [aut]
+Repository: CRAN
+Date/Publication: 2016-02-06 13:47:07
```
Comment (thomasp85) [inline DESCRIPTION]: Can you avoid changing the style and overall content of DESCRIPTION

File: R/densityClust.R [hunk 3]
```diff
@@ -69,6 +69,8 @@ NULL
 localDensity <- function(distance, dc, gaussian=FALSE) {
    # These implementations are faster by virtue of being written in C++
    # They also avoid the need to convert `distance` to a matrix. 
+   message(paste("the length of the distance:", length(distance)))
```
Comment (thomasp85) [inline R/densityClust.R]: You don't need to wrap it in paste - message does that for you

This goes for all the verbose messaging

File: R/densityClust.R [hunk 4]
```diff
@@ -150,7 +152,7 @@ estimateDc <- function(distance, neighborRateLow=0.01, neighborRateHigh=0.02) {
       # equivalent. The diagonal of the matrix will always be 0, so as long as dc 
       # is greater than 0, we add 1 for every element of the diagonal, which is
       # the same as size
-      neighborRate <- (((sum(distance < dc) * 2 + (if (0 <= dc) size)) / size - 1)) / size
+      neighborRate <- (((sum(distance < dc) * 2 + (if (0 <= dc) size)) / size - 1)) / size # sum(distance < dc) / size^2
```
Comment (thomasp85) [inline R/densityClust.R]: Can I get you to remove all non-explanatory comments from the code

File: R/densityClust.R [hunk 5]
```diff
@@ -215,15 +219,25 @@ estimateDc <- function(distance, neighborRateLow=0.01, neighborRateHigh=0.02) {
...
-        dc <- estimateDc(distance)
-    }
-    rho <- localDensity(distance, dc, gaussian=gaussian)
-    delta <- distanceToPeak(distance, rho)
-    res <- list(rho=rho, delta=delta, distance=distance, dc=dc, threshold=c(rho=NA, delta=NA), peaks=NA, clusters=NA, halo=NA)
-    class(res) <- 'densityCluster'
-    res
+densityClust <- function(distance, dc, gaussian=FALSE, verbose = F) {
+  if(missing(dc)) {
+    if(verbose)
```
Comment (thomasp85) [inline R/densityClust.R]: The style of this package is to have braces around all statements if they start on a new line

**Comment by thomasp85** [comment]
This generally looks good - If you address the comments I've made (they are all pretty minor style changes) I'll merge...

**Comment by Xiaojieqiu** [comment]
@thomasp85 thanks a lot for your time on this! I will fix those issues you mentioned (some of them results from using Rstudio) later today. Best wishes to your new transition!

**Comment by Xiaojieqiu** [comment]
@thomasp85 thanks again for your comments. I have updated the scripts. Hopefully it has resolved all your comments. Please let me know if there are any more concerns.

**Comment by thomasp85** [comment]
Can you squash the commits and then I'll merge

**Comment by Xiaojieqiu** [comment]
@thomasp85 I just learnt Squash commits. Hopefully it works fine.

**Comment by thomasp85** [comment]
Have you pushed using the --force option? It seems there are still multiple commits (the purpose of squash is to collapse all your changes into a single commit so the history doesn't get too confusing)

**Comment by Xiaojieqiu** [comment]
I forgot whether or not I used the --force option. I tried a few hours to redo the squash. But I confront a few things which I have no clue: 1. although I use **git rebase -i HEAD~10** to run, I still got the **# Rebase 7efa790..bb94145 onto 7efa790 (34 command(s))** in the file pop from this run. secondly, when I edit the file to squash all my commits into a single one as the following:   

 25 squash 57c690a size_t for integer and avoid converting dist object to matrix#
 26 squash a93e9c2 fix errors in c++ and test
 27 squash 3f21a23 update the cluster assignment
 28 squash 8acb88e add the argument peaks to findClusters.densityCluster
 29 squash ec08f8b update documentation findClusters.densityCluster
 30 squash 0014f25 prepare for pull request
 31 squash 2b58c9d prepare for pull request
 32 squash 987547d prepare for pull request
 33 squash 4c9bb94 update the Rd files
 34 squash d9fcd58 update scripts based on Thomas's suggestions
I confronted the following error: error: could not apply 9515ff3... Increase efficiency of implementation. 9515ff3 is a commit from seaaan. I got pretty frustrated by this right now. Do you have a solution for this? Otherwise could you please merge the changes  without squash the commits since it won't be essential issue and densityClust is mostly downloaded from CRAN. Let me know what is your thoughts. Thanks

**Comment by thomasp85** [comment]
Sorry for having been slow on this. I'm prepping for a CRAN update. If you're still interested in this PR will you consider merging the latest changes from master into it - then I'll merge back once all conflicts are fixed...

**Comment by Xiaojieqiu** [comment]
what is your schedule for this? I won't be available this week (probably some limited time this weekend). BTW, the KNN based approach is not compatible to the plotDensityClust function because of the implementation. I was actually think about writing a new package based on yours since this alternative KNN based method is methodological different from the original one. But I am happy to merge with your packages too. How do you think?

**Comment by thomasp85** [comment]
I don't really have a schedule - just wanted to know whether you were still interested. Is this PR fully superseded by #8? If so I'll close this and we can move the discussion there...

**Comment by Xiaojieqiu** [comment]
yes. that is right. thanks

**Pull request closed without merging**